api/src/main/java/org/apache/iceberg/actions/ExpireSnapshots.java

```diff
@@ -0,0 +1,114 @@
+import java.util.function.Consumer;
+import org.apache.iceberg.Snapshot;
+
+/**
+ * An action that expires snapshots in a table.
+ * <p>
+ * Similar to {@link org.apache.iceberg.ExpireSnapshots} but may use a query engine to distribute
+ * parts of the work.
+ */
+public interface ExpireSnapshots extends ConfigurableAction<ExpireSnapshots, ExpireSnapshots.Result> {
+  /**
+   * Expires a specific {@link Snapshot} identified by id.
+   * <p>
+   * Identical to {@link org.apache.iceberg.ExpireSnapshots#expireSnapshotId(long)}
+   *
+   * @param snapshotId id of the snapshot to expire
+   * @return this for method chaining
+   */
+  ExpireSnapshots expireSnapshotId(long snapshotId);
+
+  /**
+   * Expires all snapshots older than the given timestamp.
+   * <p>
+   * Identical to {@link org.apache.iceberg.ExpireSnapshots#expireOlderThan(long)}
+   *
+   * @param timestampMillis a long timestamp, as returned by {@link System#currentTimeMillis()}
+   * @return this for method chaining
+   */
+  ExpireSnapshots expireOlderThan(long timestampMillis);
+
+  /**
+   * Retains the most recent ancestors of the current snapshot.
+   * <p>
+   * If a snapshot would be expired because it is older than the expiration timestamp, but is one of
+   * the {@code numSnapshots} most recent ancestors of the current state, it will be retained. This
+   * will not cause snapshots explicitly identified by id from expiring.
+   * <p>
+   * Identical to {@link org.apache.iceberg.ExpireSnapshots#retainLast(int)}
+   *
+   * @param numSnapshots the number of snapshots to retain
+   * @return this for method chaining
+   */
+  ExpireSnapshots retainLast(int numSnapshots);
+
+  /**
+   * Passes an alternative delete implementation that will be used for manifests and data files.
+   * <p>
+   * Manifest files that are no longer used by valid snapshots will be deleted. Data files that were
+   * deleted by snapshots that are expired will be deleted.
+   * <p>
+   * If this method is not called, unnecessary manifests and data files will still be deleted.
+   * <p>
+   * Identical to {@link org.apache.iceberg.ExpireSnapshots#deleteWith(Consumer)}
+   *
+   * @param deleteFunc a function that will be called to delete manifests and data files
+   * @return this for method chaining
+   */
+  ExpireSnapshots deleteWith(Consumer<String> deleteFunc);
+
+  /**
+   * Passes an alternative executor service that will be used for manifests and data files deletion.
+   * <p>
+   * If this method is not called, unnecessary manifests and data files will still be deleted using
+   * a single threaded executor service.
+   * <p>
+   * Identical to {@link org.apache.iceberg.ExpireSnapshots#executeDeleteWith(ExecutorService)}
+   *
+   * @param executorService the service to use
+   * @return this for method chaining
+   */
+  ExpireSnapshots executeDeleteWith(ExecutorService executorService);
+
+  /**
+   * The action result that contains a summary of the execution.
+   */
+  interface Result {
```

Comment:
I think representing results as interfaces will allow us to evolve them easier.

Reply:
Is this a concern? I'm not sure we have had this issue yet. I don't think there is a real negative to doing this I just don't have a lot of imaginative thinking here.

Reply:
For example, `RemoveOrphanFiles` returns `List<String>` directly. It is impossible to change the return type without breaking the API. I think having a layer of abstraction would be safer. Who knows what will we come up with? What if we decide to report orphan metadata and data files separately?

Reply:
Your example should be a breaking change which is better highlighted by changing the API instead of hiding it behind an interface and depend on the documentation to describe the change.

But that is just an extreme example and I can come up with cases for both approaches, so no clean preference on my side.

Reply:
I meant the following interface:

```
interface Result {
  Iterable<String> orphanFileLocations();
}
```

Can safely evolve into (just an example, may not be the best idea):

```
interface Result {
  Iterable<String> orphanFileLocations(); // reports the same set which is a union of locations reported by methods below
  Iterable<String> orphanDataFileLocations();
  Iterable<String> orphanManifestLocations();
  Iterable<String> orphanManifestListLocations();
}
```

If we always returned just `List<String>`, we would be constrained to evolve the result type even when it is not a breaking change.

Reply:
I prefer this approach because it puts the context in the API rather than in documentation. To get the locations, it is clear that they are the `orphanFileLocations` rather than a list of strings (if you don't have Javadoc to see what the strings are).

Reply:
Resolving this. Thanks to everybody who participated.